src/sys/socket/addr.rs

```diff
@@ -1761,6 +1761,89 @@ pub mod netlink {
     }
 }
 
+#[cfg(target_os = "linux")]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+pub mod can {
+    use super::*;
+    use crate::sys::socket::addr::AddressFamily;
+    use libc::{sa_family_t, sockaddr_can, c_int, c_uint};
+    use std::{fmt, mem};
+
+    /// Address for the Linux SocketCAN implementation for the Controller Area Network
+    /// automotive bus protocol.
+    ///
+    /// # References
+    ///
+    /// [SocketCAN](https://docs.kernel.org/networking/can.html)
+    #[derive(Copy, Clone)]
+    #[allow(missing_debug_implementations)]
+    #[repr(transparent)]
+    pub struct CanAddr(pub(in super::super) sockaddr_can);
+
+    impl CanAddr {
+        /// Construct a new CAN socket address from its interface index.
+        pub fn new(ifindex: c_uint) -> Self {
+            let mut addr: sockaddr_can = unsafe { mem::zeroed() };
+            addr.can_family = AddressFamily::Can as sa_family_t;
+            addr.can_ifindex = ifindex as c_int;
+
+            Self(addr)
+        }
+
+        /// Return the socket's CAN interface index.
+        pub const fn ifindex(&self) -> c_uint {
+            self.0.can_ifindex as c_uint
+        }
+    }
+
+    impl private::SockaddrLikePriv for CanAddr {}
+    impl SockaddrLike for CanAddr {
+        unsafe fn from_raw(
+            addr: *const libc::sockaddr,
+            len: Option<libc::socklen_t>,
+        ) -> Option<Self>
+        where
+            Self: Sized,
+        {
+            if let Some(l) = len {
+                if l != mem::size_of::<libc::sockaddr_can>() as libc::socklen_t {
+                    return None;
+                }
+            }
+            if (*addr).sa_family as i32 != libc::AF_CAN {
+                return None;
+            }
+            Some(Self(ptr::read_unaligned(addr as *const _)))
+        }
+    }
+
+    impl AsRef<libc::sockaddr_can> for CanAddr {
+        fn as_ref(&self) -> &libc::sockaddr_can {
+            &self.0
+        }
+    }
+
+    impl fmt::Display for CanAddr {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "ifindex: {}", self.ifindex())
+        }
+    }
+
+    impl PartialEq for CanAddr {
```

Comment:
This implementation ignores the union field.  Does the union field really not matter?  It's usually not possible to implement `PartialEq` for anything with union fields, unless you have some way to know which field is valid.

Reply:
The most common base raw socket address does not use the union at all for anything. It just needs to specify the family (always `AF_CAN`) and the interface number. This is all that the current and next major releases of the `socketcan-rs` crate will cover.

This is all that I was thinking this `CanAddr` is meant to model.

The union is used for higher level protocols like J1939 or ISO-TP. I was thinking that those might be better modeled as separate address types like, maybe, `CanJ1939Addr`.  I don't have a really strong opinion on this, but considering that the protocols aren't typically mixed, and are implemented by separate Rust crates, having different targeted address structs at this level makes some sense.

But if we want to make this one implement all the different possibilities in one struct, I could take a stab at it.

Reply:
As an alternate, more complete implementation, we can make new functions for the additional protocols:

https://github.com/fpagliughi/nix/blob/88df2c37b54a7444257d8f3ae7ad928e3bea4aea/src/sys/socket/addr.rs#L1784-L1808

Then we could implement `PartialEq` as a bitwise binary comparison of the whole struct, and `Hash` by considering the struct similarly as a slice of bytes over the size of the underlying sockaddr struct.

https://github.com/fpagliughi/nix/blob/88df2c37b54a7444257d8f3ae7ad928e3bea4aea/src/sys/socket/addr.rs#L1849-L1879

This works insofar as two addresses created with the same parameters to any of the `new...()` functions will report as equal, and any two addresses that are equal will produce the same hash.

There is the argument that addresses created for the different protocols could, with certain parameters, produce the same binary address value, even though they were logically different as they were meant for different protocols. If that argument were to prevail, I think it would indicate that these are actually different address structures and should be different types.

I don't have a strong opinion. I could go either way. They both work equally well for my use case.

Reply:
I agree this isn't a general way to compare unions. But all `sockaddr` types are like one big de-facto union in C, and the typical convention has always been to zero out the whole struct. This is enforced here with the `new...()` functions, but since users can create these types from underlying structs, they can bypass the conventions and create incompatible values.

But what bothers me more about this implementation is that users can compare, say, a J1939 address to a raw address or an ISO-TP address. These really aren't the same thing at all. They're actually different protocols.

So it makes me think these should be different types. So we could do something like this (yet another implementation):

https://github.com/fpagliughi/nix/blob/2276af8daa455a4e7762617872c40248fc629d8f/src/sys/socket/addr.rs#L1764-L2111

It's a bit verbose, and I wonder if there isn't some way to consolidate the common code between them. But it's clean in keeping the types separate, and the `PartialEq` and `Hash` implementations should be valid for each separate type.

Reply:
If you split the type into three, how would `SockaddrLike::from_raw` work?